### PR TITLE
Use assembly routines for Dot and AddScaled*. Add benchmark for AddScale...

### PR DESCRIPTION
**Pull request by btracey**
...dTo

### CODE REVIEW

**Comment by btracey** [comment]
The change to AddScaledTo is not a clear win. 

brendan:~/Documents/mygo/src/github.com/gonum/floats$ benchcmp orig.bench new.bench 
benchmark                      old ns/op     new ns/op     delta       
BenchmarkAddScaledToSmall      20.0          21.2          +6.00%      
BenchmarkAddScaledToMedium     1249          525           -57.97%     
BenchmarkAddScaledToLarge      125598        105003        -16.40%     
BenchmarkAddScaledToHuge       21816250      27014749      +23.83%  

However, the change is minor for small entries, very significantly positive for medium sizes, and significantly negative for Huge. It seems to me this is a win because things will usually be smaller, and I've seen a number of PRs over time to improve Copy, and if/when we have AVX the assembly routine will be faster. Both of these should speed up Huge and make it a net win. If not, we could add a size length switch of some kind.

File: floats.go [hunk 1]
```diff
@@ -69,9 +69,8 @@ func AddScaledTo(dst, y []float64, alpha float64, s []float64) []float64 {
 	if len(dst) != len(s) || len(dst) != len(y) {
 		panic("floats: lengths of slices do not match")
 	}
-	for i, val := range s {
-		dst[i] = y[i] + alpha*val
-	}
+	copy(dst, y)
```
Comment (fhs) [inline floats.go]: There's no reason to have a copy here. I can change DaxpyUnitary to take a destination argument: `DaxpyUnitary(alpha, x, y, z)` where `z[i] = alpha*x[i] + y[i]`

Comment (btracey) [inline floats.go]: Is it too much to have both? The current version shows up constantly in
blas,  whereas the three element argument just shows up here (and lots of
code that calls this,  especially in optimize).

Comment (fhs) [inline floats.go]: If I'm understanding you correctly, there is no reason to have two versions of DaxpyUnitary. The blas/native code will call it as `DaxpyUnitary(alpha, x, y, y)` and the code here will call it as `DaxpyUnitary(alpha, x, y, z)`.

Comment (btracey) [inline floats.go]: I assume there's no performance penalty?

Do you have a preference @kortschak?

Comment (fhs) [inline floats.go]: There's no performance penalty in the assembly version. I haven't checked the pure Go version, where another bound check is introduced in z[i] when you convert
```y[i] += alpha * v``` to ```z[i] = alpha * v + y[i]```.

Comment (kortschak) [inline floats.go]: A single routine is preferable IMO.

**Comment by btracey** [comment]
SGTM

If you would like to make that change,  I'll update this and the blas PRs. Thanks for the input,  and sorry if I'm being overly cautious.

**Comment by btracey** [comment]
PTAL

**Comment by fhs** [comment]
What's the benchmark for AddScaledTo now? Did it improve?

**Comment by btracey** [comment]
Much better, thank you

benchmark                      old ns/op     new ns/op     delta       
BenchmarkAddScaledToSmall      20.9          14.0          -33.01%     
BenchmarkAddScaledToMedium     1219          318           -73.91%     
BenchmarkAddScaledToLarge      120523        70862         -41.20%     
BenchmarkAddScaledToHuge       19079937      19094861      +0.08%

**Comment by fhs** [comment]
LGTM

The benchmarks can be made more compact by calling RandomSlice from within benchmarkAddScaledTo but I'll leave it up to you.

**Comment by btracey** [comment]
I'm going to submit as is and make a PR to fix all the benchmarks.

**Pull request merged**